### PR TITLE
Fix kube-addon-manager overwriting resources with EnsureExists set

### PR DESCRIPTION
**Pull request by spencer-p**
The addon manager readme states

> - Addons with label `addonmanager.kubernetes.io/mode=EnsureExists` will be checked for
> existence only. Users can edit these addons as they want.

However, the start_addon function was using `kubectl apply` to create resources regardless of mode. Users who change EnsureExists resources have their configs wiped out and replaced when the addon manager restarts.

This change switches between `kubectl create` and `kubectl apply` depending on the addon's mode.

Noting this fix may have to be re-implemented after #92978.

/kind bug

```release-note
NONE
```

### CODE REVIEW

**Comment by MrHohn** [comment]
@spencer-p It looks like the `start_addon` func, in which `kubectl apply` is used regardless of mode, is only triggered for manifests stored in `/etc/kubernetes/admission-controls`. Is extending the EnsureExists/Reconcile logic to cover admission-control objects the intention here?

**Comment by spencer-p** [comment]
@mrhohn Yes. This would affect `cluster/gce/addons/limit-range/limit-range.yaml` in particular.

**Comment by spencer-p** [comment]
polite bump - review greatly appreciated as this issue has moderately high priority for users

**Comment by MrHohn** [comment]
(Looks like the /assign command failed?)
/assign

**Comment by spencer-p** [comment]
Looks like prow jumped the gun on that commit. Looks like a lot of container pull errors, likely because I hadn't pushed images to staging.
/retest

**Comment by MrHohn** [comment]
> Looks like prow jumped the gun on that commit. Looks like a lot of container pull errors, likely because I hadn't pushed images to staging.

Yep, we generally do this in three steps:
1. PR to edit the script.
2. Publish a new image.
3. PR to bump the image version.

**Comment by spencer-p** [comment]
/retest
Looks like the previous tests failed due to an SSL certificate and what seems to be a prow bug (test failed but 37/37 passed??)

**Review by MrHohn — APPROVED** [review]
Thanks for the fix @spencer-p! I don't have much comment - looks like reasonable changes to me. Have you built a test image and run it on a test cluster to be sure? (Very easy to make mistake when dealing with shell script..)

/approve

**Comment by spencer-p** [comment]
Thanks @MrHohn. I have tested these changes in a cluster.

Here's the manual tests I've done. Assuming you're logged into the leader master node and the `LimitRange` config is in `/etc/kubernetes/admission-controls`:

### Creates missing resource
1. `kubectl delete limits/limits`
2. `crictl ps | grep addon | awk '{print $1}' | xargs crictl stop` (this command will kill and restart the addon-manager)
3. `kubectl get limits/limits -oyaml` should return the object

### Does not re-create existing resource
1. `EDITOR="sed -i 's/100m/600m/'" kubectl edit limits/limits`
2. Restart the addon manager
3. `kubectl get limits/limits -oyaml` should reflect the 600m.
4.  Checking the kube addon manager logs should show skipping the resource.

And then editing `/etc/kubernetes/admission-controls/limit-range/limit-range.yaml` to have `=Reconcile` instead of `=EnsureExists`. Then the results of the first test should be identical, but the second test should overwrite it.

These aren't complicated tests (though they are somewhat time based, esp. when the expected result is "no change").

So if anyone has thought or guidance on adding E2E tests, it would be much appreciated.

**Comment by MrHohn** [comment]
Adding Jake as he is re-implementing this.
/cc @dekkagaijin

There is an e2e test that exercises addon manager functionalities: https://github.com/kubernetes/kubernetes/blob/master/test/e2e/cloud/gcp/addon_update.go ([dashboard](https://k8s-testgrid.appspot.com/google-gce#gce-cos-master-slow&include-filter-by-regex=Addon))

However this new logic only applies to start up time so might be trickier to set up the test. I'm fine with the manual tests you have done.

Could you squash the follow-up commits? Thanks!

**Review by dekkagaijin — APPROVED** [review]
/lgtm

thanks for the heads up

**Comment by MrHohn** [comment]
Thanks!
/lgtm

**Comment by dekkagaijin** [comment]
I'll implement these changes in the [Golang port](https://github.com/kubernetes-sigs/cluster-addons/pull/76) when both PRs get approved

File: cluster/addons/addon-manager/kube-addons.sh
Comment (liggitt) [inline cluster/addons/addon-manager/kube-addons.sh:164]: /hold 

does this work if the string contains multiple objects, some with EnsureExists and some without?

why is a different approach being taken here than in ensure_addons/reconcile_addons (which use label selectors)

Comment (spencer-p) [reply cluster/addons/addon-manager/kube-addons.sh:164]: It wouldn't work, good point. There don't seem to be any multi-object files in testing.

I seemed to be under the impression that the label selectors wouldn't work as intended previously, and I don't have any notes on why I thought that was the case. I will note that this image is currently bundled with kubectl version 1.13.2.

I'll test a change with separate `kubectl apply`/`kubectl create` commands with label selectors.

**Comment by spencer-p** [comment]
I've made some significant changes to account for @liggitt 's concerns about strings with multiple resources and label selectors.

In summary:
 - `kube-addons.sh` now only contains functions and definitions.
 - `kube-addons-main.sh` now contains the main entry point for the container and all the code removed from `kube-addons.sh`
 - `kube-addons-test.sh` is a script that runs integration tests for the affected portions of the codebase.
 - You can run the tests with `make test`, and they require a cluster to be set up and configured for `kubectl`. Using `make-test` will use the same kubectl version that will be packaged in the final image.
 - I factored out functions from `create_resource_from_string` to handle the subsets of EnsureExists resources and Reconcile resources. `create_resource_from_string` only succeeds when both inner functions succeed.

**Comment by liggitt** [comment]
Thanks for the update, I don't have review bandwidth for this at the moment, but if it is consistent with the handling of labels on multiple resources, my objections are resolved.

I will note that a significant refactor/rewrite of bash scripts require exceptionally careful review to ensure equivalent behavior of pieces that were not intended to change. Once approvers/reviewers are satisfied this is correct, they can remove the hold at will.

File: cluster/addons/addon-manager/kube-addons-main.sh
Comment (MrHohn) [inline cluster/addons/addon-manager/kube-addons-main.sh:32]: Just double checking - I assumed nothing changed after this line?

File: cluster/addons/addon-manager/kube-addons-main.sh
Comment (MrHohn) [inline cluster/addons/addon-manager/kube-addons-main.sh:57]: Instead of figuring out every single yaml/json file and reconcile/ensure them one by one, I have been wondering whether a simple `kubectl apply/create -f` would work? My understanding is the command automatically ignores unrelated formats?

There are probably hidden complexity in the retry logic during master startup, please feel free to point out :)

File: cluster/addons/addon-manager/kube-addons.sh
Comment (MrHohn) [inline cluster/addons/addon-manager/kube-addons.sh:182]: Would be great to log this slightly differently so that we can distinguish ensure and reconcile? e.g. `Successfully reconciled ...`

File: cluster/addons/addon-manager/kube-addons-test.sh
Comment (MrHohn) [inline cluster/addons/addon-manager/kube-addons-test.sh:24]: Remove this line?

File: cluster/addons/addon-manager/kube-addons-test.sh
Comment (MrHohn) [inline cluster/addons/addon-manager/kube-addons-test.sh:96]: Add a step that make sure it is 50m now?

File: cluster/addons/addon-manager/kube-addons-test.sh
Comment (MrHohn) [inline cluster/addons/addon-manager/kube-addons-test.sh:63]: I find it easy to follow the exact test steps when reading the codes, but not as easy when executing it. Could we log necessary info about what the test is doing during the process?

File: cluster/addons/addon-manager/kube-addons-test.sh
Comment (MrHohn) [inline cluster/addons/addon-manager/kube-addons-test.sh:249]: nit: name the variable something like `test_func` to be clear?

File: cluster/addons/addon-manager/kube-addons-test.sh
Comment (MrHohn) [inline cluster/addons/addon-manager/kube-addons-test.sh:255]: nit: call `teardown` after the if block?

**Review by MrHohn — COMMENTED** [review]
Left some comments as the first pass, thanks for adding the test!

File: cluster/addons/addon-manager/kube-addons.sh
Comment (spencer-p) [reply cluster/addons/addon-manager/kube-addons.sh:171]: I'll rename these. I'll also note that the lack of `--prune` is not a regression - the original `create_resource_from_string` does not prune.

File: cluster/addons/addon-manager/kube-addons.sh
Comment (spencer-p) [reply cluster/addons/addon-manager/kube-addons.sh:60]: I actually added the export because shellcheck noticed that this variable wasn't used in this file, and so should either be deleted or exported. The export doesn't have any effect other than making the variable available to sub-processes, which shouldn't have any effect here.

It might be more readable to just `# shellcheck disable` for those variables.

File: cluster/addons/addon-manager/kube-addons-main.sh
Comment (spencer-p) [reply cluster/addons/addon-manager/kube-addons-main.sh:32]: Correct :)

File: cluster/addons/addon-manager/kube-addons-main.sh
Comment (spencer-p) [reply cluster/addons/addon-manager/kube-addons-main.sh:57]: I have no idea why this was written with a file walk. I think this one is out of scope for this PR.

**Review by MrHohn — APPROVED** [review]
Thanks! I don't spot any issue, and now it seems to handle the addon-manager labels in the same way as the rest of the script.

Could you squash the fix up commits?

**Comment by spencer-p** [comment]
Rebase is done 👍

**Comment by MrHohn** [comment]
Thanks!
/lgtm
/hold cancel
Un-holding this to move forward. After bumping the manifest, lets monitor the e2e tests (e.g. https://k8s-testgrid.appspot.com/google-gce#gce-cos-master-slow&include-filter-by-regex=Addon) closely to double check the behavior.

**Comment by liggitt** [comment]
were the 9.1.2 images ever used? the manifest still shows 9.1.1 at HEAD:

https://github.com/kubernetes/kubernetes/blob/f48972e671e5f338d0d6adfe6055aba4534569ad/cluster/gce/manifests/kube-addon-manager.yaml#L26

**Comment by dekkagaijin** [comment]
@liggitt doesn't seem so. I'm still planning on integrating this fix into the golang re-implementation and releasing that soonish

**Pull request merged**
